src/solvers/dgsem_tree/dg_2d.jl

```diff
@@ -1036,6 +1097,148 @@ function calc_mortar_flux!(surface_flux_values,
+                               cache,
+                               u_large::AbstractArray{<:Any, 2},
+                               u_large_upper::AbstractArray{<:Any, 2},
+                               u_large_lower::AbstractArray{<:Any, 2},
+                               direction, large_element_id, mortar,
+                               fstar_upper_correction, fstar_lower_correction,
+                               surface_flux)
+    orientation = cache.mortars.orientations[mortar]
+
+    for j in eachnode(dg)
+        u_rr_upper = get_node_vars(u_large_upper, equations, dg, j)
+        u_rr_lower = get_node_vars(u_large_lower, equations, dg, j)
+        for i in eachnode(dg)
+            u_ll_large = get_node_vars(u_large, equations, dg, i)
+
+            # Call pointwise two-point numerical flux function
+            flux_upper = surface_flux(u_ll_large, u_rr_upper,
+                                      orientation, equations)
+
+            flux_lower = surface_flux(u_ll_large, u_rr_lower,
+                                      orientation, equations)
+
+            # Copy flux back to actual flux array
+            set_node_vars!(fstar_upper_correction, flux_upper, equations, dg, i, j)
+            set_node_vars!(fstar_lower_correction, flux_lower, equations, dg, i, j)
+        end
+    end
+
+    # Loop over all nodes on large face
+    for i in eachnode(dg)
+        # Loop over all variables
+        for v in eachvariable(equations)
+            # Calculate flux corrections for i'th node
+            flux_correction = zero(eltype(fstar_upper_correction))
+
+            # Loop over inner-element nodes (see Fig. 3) of
+            #   Chan, Bencomo, Del Rey Fernández (2021).
+            #   Mortar-based Entropy-Stable Discontinuous Galerkin Methods on 
+            #     Non-conforming Quadrilateral and Hexahedral Meshes.
+            #   https://doi.org/10.1007/s10915-021-01652-3
+            for j in eachnode(dg)
+                # j-local flux: "Forward" flux
+                f_j_upper = fstar_upper_correction[v, i, j]
+                f_j_lower = fstar_lower_correction[v, i, j]
+
+                # Subtract "forward" flux
+                for k in eachnode(dg)
+                    f_j_upper -= mortar_ec.forward_upper[j, k] *
+                                 fstar_upper_correction[v, k, j]
+                    f_j_lower -= mortar_ec.forward_lower[j, k] *
+                                 fstar_lower_correction[v, k, j]
+                end
+
+                # Add to flux correction involving "reverse" flux
+                flux_correction += mortar_ec.reverse_upper[i, j] * f_j_upper
+                flux_correction += mortar_ec.reverse_lower[i, j] * f_j_lower
+            end
+
+            # Finally, add to surface flux values
+            surface_flux_values[v, i, direction, large_element_id] += flux_correction
+        end
+    end
+end
+
+function calc_mortar_flux!(surface_flux_values, u,
```

Comment:
I'm not too familiar with this part of Trixi; is it mostly the same as `calc_mortar_flux!` for the standard L2 mortars except for the correction call at the end?

Reply:
Yes exactly :+1: